tensorboard/plugins/graph/README.md

```diff
@@ -121,9 +121,384 @@ describing the motivations, research, and process behind it.
+            later" (TODO: why?).
+    -   let templates = groupTemplateAndAssignId(nnGroups, verifyTemplate);
+    -   For each MetaNode in a nnGroup
+
+        -   For all previous clusters, try to add the current MetaNode to the
+            previous cluster if their metagraphs are similar. Note: this likely
+            results in a quadratic+ time complexity.
+        -   MetaGraph similarity conditions
+
+            -   Must have same sorted "degree sequence" overall nodes
+            -   Must have same # of source nodes
+            -   When performing a DFS nodes simultaneously in both graphs using
+                2 pointers, one in each MetaGraph, starting from the sources and
+                traversing successors sorted by (op, templateId, neighborCount,
+                predecessorCount, successorCount, nodeName), the MetaGraphs
+                differ:
+                -   If the nodes differ by Op, templateId, type, seriesCount
+                -   If the number of ousorted list of successors (nodes at
+                    outgoing edges) differ
+                -   If one MetaGraph revisits a node while the other MetaGraph
+                    does not
+
+        -   Generate a clusterId as `${signature}[${clusterIndex}]`.
+
+            -   Since 2 MetaNodes might have the same signature, but have
+                different internal structure, they may end up in different
+                clusters, e.g.
+
+                -   MetaNode signature "depth=1 |V|=3 |E|=0 [ops] Add=3"
+                -   MetaNode.templateId `signature + '[' + clusters.length +
+                    ']'`
+
+        -   Treat the ids from the final clusters as 'templateId's.
+
+-   Returns
+
+    ```
+    {
+      root: Metanode;
+      libraryFunctions: {
+        [key: string]: LibraryFunctionData;
+      };
+      templates: {
+        "depth=1 |V|=3 |E|=0 [ops] Add=3[0]": {level: 0, nodes: MetaNode[]}
+      } as {[templateId: string]: string[]};
+      /** List of all device names */
+      devices: string[];
+      /** List of all XLA cluster names */
+      xlaClusters: string[];
+      /** True if at least one tensor in the graph has shape information */
+      hasShapeInfo: boolean;
+      /** The maximum size across all meta edges. Used for scaling thickness. */
+      maxMetaEdgeSize: number;
+      getBridgegraph(nodeName: string): graphlib.Graph;
+      ...
+    }
+    ```
+
+### Rendering the view
+
+TODO
+
+### Node extraction
+
+TODO
+
+## Glossary
+
+Communicating the inner workings of the Graph dashboard requires introducing
+some dashboard-specific concepts used throughout the web frontend code. Much of
+this terminology comes from
+[existing literature](https://www.cs.ubc.ca/labs/imager/tr/2008/Archambault_GrouseFlocks_TVCG/grouseFlocksSub.pdf).
+
+-   **GraphDef**: A 'base graph' structure, loosely tied to TensorFlow's
+    [graph.proto](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/graph.proto).
```

Comment:
You don't like the TB version? :D https://github.com/tensorflow/tensorboard/blob/master/tensorboard/compat/proto/graph.proto